Make @OpenShiftResource annotation work again with plain text

goldmann opened the pull request:
This proposal uses a temporary file workaround for the #1045 issue.

Fixes #1045

lordofthejars commented:
Thinking about code, I am not sure if the best way is to do this or just create an interface `ResourceContent` which has a method called `getContent`.

Then we can have an implementation called `URLResourceContent` and `TextResourceContent`. The first one just do all `BufferReading` thingy to get an `String` from `URL` meanwhile the other one is just returning the String.

@dipak-pawar any thoughts about it?

dipak-pawar commented:
@lordofthejars +1 for the suggestion which you proposed in above comment.

lordofthejars commented on kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java:
```diff
@@ -24,7 +27,20 @@ public static URL resolve(String location) {
+                // In case it is not a http, file nor a classpath protocol,
+                // we assume that this is plain text. We store it in a temporary
+                // file and return the URL to it.
+                File tmp = File.createTempFile("arquillian-cube", ".res");
```
Can you move the creation of temporal file into a method like `createTemporalDefinition` which returns a `File` that is used later to create the URI? Just to have method more compact.

goldmann commented:
@lordofthejars How about this?

dipak-pawar commented:
@goldmann https://travis-ci.org/arquillian/arquillian-cube/jobs/388249666#L1213 Can you check this?

lordofthejars commented:
@goldmann @dipak-pawar Are you receiving this compilation error in local too?

dipak-pawar commented:
yes @lordofthejars, I am getting this locally too.

dipak-pawar commented on kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resolver/ResourceResolver.java:
```diff
@@ -9,22 +11,22 @@
     static final String CLASSPATH_PREFIX = "classpath:";
 
-    public static URL resolve(String location) {
+    public static InputStream resolve(String content) {
```
This method is used in https://github.com/arquillian/arquillian-cube/blob/master/kubernetes/kubernetes/src/main/java/org/arquillian/cube/kubernetes/impl/resources/KubernetesResourcesApplier.java#L102 which is expecting URL. You have to modify it to make it working as per modification

goldmann commented:
Ooops, this one should be fixed now :)

goldmann commented:
It looks like I'm reading some resources too early, hold on tight.

goldmann commented on openshift/openshift/src/main/java/org/arquillian/cube/openshift/impl/utils/TemplateUtils.java:
```diff
@@ -72,7 +75,12 @@ public static String readTemplateUrl(Template template, CubeOpenShiftConfigurati
+                templateUrl = new URL(url).toString();
+            } catch (MalformedURLException ex) {
+                throw new IllegalArgumentException(ex);
+            }
```
The only reason we were "resolving" it earlier was to actually wrap it in URL and return string representation of it. Now we return `InputStream` in the `resolve` method reading it earlier. I decided to just validate the URL in the old form, as you can see.

goldmann commented:
I'm giving up today, will start again tomorrow, forgot that it can be other things than url too :)

goldmann commented:
Tests are failing today badly, because of timeout issues :(

goldmann commented:
Phew, finally! I reverted to using the temporary file to return URL because returning `InputStream` was too problematic. I extracted preparing the file to separate method, as suggested.
